test/dataloader2/test_dataloader2.py

```diff
@@ -352,6 +352,21 @@ def clean_me(process, req_queue, res_queue):
         clean_me(process, req_queue, res_queue)
 
 
+class NonShardableDataPipe(IterDataPipe):
```

Comment:
> BTW why do we want to use this sharding as default? seems so expensive (e.g. reading/decompressing data times and discarding most of them?...)

I think it's more OSS use case where users want to provide a list of file paths. When MP takes place, with the explicit `sharding_filter`, we only do read/decompress partial dataset. This requires users to place `sharding_filter` to the right place.

> I am wondering if we should give more specific name? Such as "filter_shard" vs. "roundrobin_shard" 

I see. We can let users to define `roundrobin_shard` explicitly (alternative of `sharding_filter`). Then, we find lowest common ancestor  of those `roundrobin_shard` in the graph. This is actually a great idea! Thank you.
Then, we don't necessarily bother to define the non-shardable DataPipe. As long as `roundrobin_shard` is presented, we would treat the prior graph as the non-sharding DataPipes.

Reply:
There is one more thing that needs to carefully design. When `sharding_filter` is presented, we will do distributed sharding first then do mp sharding. However, it becomes more complicated for `sharding_roundrobin`. There are two (three) different scenarios:
- If we don't want to do distributed sharding when the backend service has already done that, it's quite straight forward to launch one non-sharding process per distributed node. Proposed API: `sharding_roundrobin(sharding_group_filter=SHARDING_PRIORITIES.MULTIPROCESSING)`
- It becomes more complicated if we want to do distributed sharding as well, because distributed sharding can be achieved by two different ways.
  - Do distributed sharding like a `filter` (emit `world_size-1/world_size`) and run a non-sharding process to dispatch data from this process to worker processes on the same node. Proposed API:  `sharding_filter(sharding_group_filter=SHARDING_PRIORITIES.DISTRIBUTED).sharding_roundrobin(sharding_group_filter=SHARDING_PRIORITIES.MULTIPROCESSING)`.
  - Do a global round-robin distributed sharding. In that case, we can launch one process on rank 0 and distribute data to other distributed workers. Proposed API:  `sharding_roundrobin()` equivalent to `sharding_roundrobin(sharding_group_filter=(SHARDING_PRIORITIES.DISTRIBUTED, SHARDING_PRIORITIES.MULTIPROCESSING))`

This requires some changes to [`apply_sharding`](https://github.com/pytorch/pytorch/blob/d6efd25d1e51337f94229036825810dc09c80824/torch/utils/data/graph_settings.py#L35-L61) to accept multiple sharding in a branch, as long as they have exclusive `sharding_priorities`.